org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/utils/MarkerUtils.java

```diff
@@ -0,0 +1,171 @@
+import org.eclipse.tm4e.ui.TMUIPlugin;
+import org.eclipse.tm4e.ui.internal.model.TMDocumentModel;
+
+public final class MarkerUtils {
+
+	private abstract static class AbstractMarkerConfig {
+		final String type;
+		final int priority;
+		final int severity;
+
+		AbstractMarkerConfig(final String type, final int priority, final int severity) {
+			this.type = type;
+			this.priority = priority;
+			this.severity = severity;
+		}
+	}
+
+	private static final class ProblemMarkerConfig extends AbstractMarkerConfig {
+		ProblemMarkerConfig(final int severity) {
+			super(PROBLEMMARKER_TYPE, IMarker.PRIORITY_NORMAL, severity);
+		}
+	}
+
+	private static final class TaskMarkerConfig extends AbstractMarkerConfig {
+		TaskMarkerConfig(final int priority) {
+			super(TASKMARKER_TYPE, priority, IMarker.SEVERITY_INFO);
+		}
+	}
+
+	private static final String TEXTMARKER_TYPE = "org.eclipse.tm4e.ui.textmarker";
+	private static final String PROBLEMMARKER_TYPE = "org.eclipse.tm4e.ui.problemmarker";
+	private static final String TASKMARKER_TYPE = "org.eclipse.tm4e.ui.taskmarker";
+
+	private static final Map<String, AbstractMarkerConfig> MARKERCONIG_BY_TAG = Map.of(
+			// problem markers:
+			"BUG", new ProblemMarkerConfig(IMarker.SEVERITY_ERROR),
+			"NOTE", new ProblemMarkerConfig(IMarker.SEVERITY_INFO),
+			// task markers:
+			"FIXME", new TaskMarkerConfig(IMarker.PRIORITY_HIGH),
+			"HACK", new TaskMarkerConfig(IMarker.PRIORITY_NORMAL),
+			"TODO", new TaskMarkerConfig(IMarker.PRIORITY_NORMAL),
+			"XXX", new TaskMarkerConfig(IMarker.PRIORITY_NORMAL));
+
+	private static final Pattern TAG_SELECTOR_PATTERN = Pattern.compile(
+			"\\b(" + MARKERCONIG_BY_TAG.keySet().stream().collect(Collectors.joining("|")) + ")\\b");
+
+	private static int getLineNumber(final IMarker marker) {
+		try {
+			final var lineNumberAttr = marker.getAttribute(IMarker.LINE_NUMBER);
+			if (lineNumberAttr instanceof final Integer lineNumber)
+				return lineNumber;
+		} catch (final CoreException ex) {
+			TMUIPlugin.log(new Status(IStatus.ERROR, TMUIPlugin.PLUGIN_ID, ex.getMessage(), ex));
+		}
+		return -1;
+	}
+
+	/**
+	 * Updates all TM4E text markers of the corresponding document starting from
+	 * <code>event.ranges.get(0).fromLineNumber</code> until the end of the document.
+	 */
+	public static void updateTextMarkers(final ModelTokensChangedEvent event) {
+		final ITMModel model = event.model;
+		if (model instanceof final TMDocumentModel docModel) {
+			try {
+				updateTextMarkers(docModel, event.ranges.get(0).fromLineNumber);
+			} catch (final CoreException ex) {
+				TMUIPlugin.log(new Status(IStatus.ERROR, TMUIPlugin.PLUGIN_ID, ex.getMessage(), ex));
+			}
+		}
+	}
+
+	/**
+	 * Updates all TM4E text markers of the {@link TMDocumentModel}'s document starting from the given line number until
+	 * the end of the document.
+	 *
+	 * @param startLineNumber 1-based
+	 */
+	public static void updateTextMarkers(final TMDocumentModel docModel, final int startLineNumber)
```

Comment:
I think you should update some existing marker and not delete all and recreate. I suggest to see the update marker of LSP4E at https://github.com/eclipse/lsp4e/blob/ca172adc3afec075357986c6c8c5459335e0e98a/org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/diagnostics/LSPDiagnosticsToMarkers.java#L154

Reply:
I now implementing a diffing mechanism.